File: 0784-letter-case-permutation/0784-letter-case-permutation.py
class Solution(object):
    def swapcase(self, data, idx) :
        if data[idx].islower() :
            data[idx] = data[idx].upper()
        else :
            data[idx] = data[idx].lower()
    def permute(self, data, i, length, result):
        if i == length :
            result.append(''.join(data))
        else :
            # Include the current character as is
            self.permute(data, i + 1, length, result)

            if data[i].isalpha() :
                self.swapcase(data, i)
                self.permute(data, i+1, length, result)
                self.swapcase(data, i)
                
        return list(set(result))          
    def letterCasePermutation(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        if len(s) == 1 :
            if s[0].isalpha() :
                return [s.upper(), s.lower()]
        ans = self.permute(list(s), 0, len(s), [])
        return ans

File: 0784-letter-case-permutation/0784-letter-case-permutation.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_set>

class Solution {
public:
    void swapcase(std::string &data, int idx) {
        if (islower(data[idx])) {
            data[idx] = toupper(data[idx]);
        } else {
            data[idx] = tolower(data[idx]);
        }
    }

    void permute(std::string &data, int i, int length, std::unordered_set<std::string> &result) {
        if (i == length) {
            result.insert(data);
        } else {
            // Include the current character as is
            permute(data, i + 1, length, result);

            if (isalpha(data[i])) {
                swapcase(data, i);
                permute(data, i + 1, length, result);
                swapcase(data, i); // Swap back
            }
        }
    }

    std::vector<std::string> letterCasePermutation(std::string s) {
        std::vector<std::string> ans;
        std::unordered_set<std::string> resultSet;

        if (s.length() == 1) {
            if (isalpha(s[0])) {
                std::string upperCase(1, toupper(s[0]));
                std::string lowerCase(1, tolower(s[0]));
                return {upperCase, lowerCase};
            }
        }

        permute(s, 0, s.length(), resultSet);
        ans.assign(resultSet.begin(), resultSet.end());
        return ans;
    }
};
